### PR TITLE
Added OpenIOC and MISP xml based automated attribute population

### PR DESCRIPTION
**Pull request by KennethAdamMiller**
So far, populating attributes through the API is a manual process, with the only exception being an out of date example demonstrating how to map IOC xml attributes to MISP xml attributes. MISP's web UI has an interface to allow users to upload their OpenIOC xml files for importing, but this same functionality is not replicated in the PyMISP apis.

I needed this functionality, so I tested the examples, and then tested the new apis after I added them to the PyMISP api.
For each IOC or MISP xml api offered, the user can either pass in a filename or a BytesIO object that can be manipulated with read. This prevents the user who has the xml information in memory from having to needlessly save it to disk in the case that an open call is made within the api by splitting the interpretation of the xml from the input type requirements.

One new dependency is BeautifulSoup 4. Likely the example ioc2misp.py could be updated, but I have what I need.

### CODE REVIEW

File: pymisp/api.py [hunk 1]
```diff
@@ -700,12 +763,51 @@ def upload_samplelist(self, filepaths, event_id, distribution, to_ids, category,
         to_post['request']['files'] = files
         return self._upload_sample(to_post)
 
+    def upload_misp_report(self, event, mispreport_file):
+        with open(mispreport, "r") as misp_report_file:
+            return self.attach_misp_attributes(event, misp_report_file.read())
+
+    def attach_misp_attributes(self, event, misp_xml):
```
Comment (Rafiot) [inline pymisp/api.py]: Just to make sure I get this function right: you get the XML export of a MISP event, extract all the attributes with BS4 and add them back to an other event, right? Why not using a MISP Json output directly?

**Comment by Rafiot** [comment]
What is the difference with this script at the end: https://github.com/CIRCL/PyMISP/tree/master/examples/ioc-2-misp

File: pymisp/api.py (hunk 1 above)
Comment (KennethAdamMiller) [inline pymisp/api.py]: I didn't see that you had commented on this until just now.

Because I'm not pulling attributes from MISP only to add them back to MISP. I'm pulling MISP and indicator content from other analysis utilities and then uploading them to a MISP server.

**Comment by KennethAdamMiller** [comment]
Possibly the fact that that script isn't part of the APIs in order that I can make what I do clean and properly organized. There's a value in having the API have that particular function in that I don't have to call a dedicated script in order to get something that is part of a pipeline. It should just be a function.

If you want, I can go back over that script that you pointed out with the new API and reduce it's size. It appears more organized to me to put them where I have.

**Comment by KennethAdamMiller** [comment]
@Rafiot Please mention me in your replies so that my gmail route the email to the correct category.

**Comment by KennethAdamMiller** [comment]
@Rafiot I understand you're busy; do you fundamentally disagree with the organization that I've proposed? 

We use Joe Security Sandbox where I work, and it generates MISP reports, so that we have to collect them up and process them into a MISP server. It's rather nasty reaching into the examples or keeping in my code, when instead the examples could just be refactored to use new exposed code in the api. Yes, no?

**Comment by Rafiot** [comment]
I had no time to look at your code in the last few weeks, I'll do it asap.

**Comment by KennethAdamMiller** [comment]
@Rafiot Oh, I didn't see that you had looked at my code. I understand that you're busy. 

1) Ok, I understand that.

2) I think that has been asked already. If you wanted to access that functionality as part of PyMISP, it isn't part of the module. I am a module consumer, and I believe that ioc-2-misp is useful functionality. I offered somewhere to refactor ioc-2-misp to use the new library code, but I don't know that I ever got a response. Yes, I did take and refactor that into the api, but it's useful to have it there. In my opinion, if I can't access ioc-2-misp from a function standpoint, as in call this function, ok it is done now, then it's not what I want to use. Using ioc-2-misp as a script, which is what it is, isn't suitable to our use case.

2.1) Ah, so then I could split the api functionality of uploading it to MISP into a reusable piece, and refactor the conversion out into some interface, and then implement it once with this code you are talking about. Correct?

3) Well it's in the function name what the expected content format is; I don't know why it is "generic" to you. But I do agree with 2.1, so I can make those changes with no fuss.

**Comment by Rafiot** [comment]
I'm working on adding tools to the API without disturbing too much the current code base, the initial work is there: https://github.com/CIRCL/PyMISP/pull/68

Do you think you can adapt your code in a similar fashion?

**Comment by Rafiot** [comment]
@KennethAdamMiller This is the way I'd like to implement your PR upstream: https://github.com/MISP/PyMISP/blob/tooling/pymisp/tools/openioc.py

Would it be fine for you?

**Comment by Rafiot** [comment]
I merged the other implementation upstream, please to a PR on this one if needed.

**Comment by KennethAdamMiller** [comment]
@Rafiot Sorry for the slow response, I have been extremely busy. 

That's so great you've found the time to break working code that was implemented, tested and working on our end for the sake of being different, especially since the needed functions are named different, aren't exposed via the api, and use different signatures. None of your code base is disturbed; we just added functions to the API. If it had bothered you so much, couldn't we just modify what I had a bit - the same function signatures and types, but moved to respective modules or something? If it were that way, I could just modify my function call, or import some module beneath pymisp. I just want my code to work and to focus on what matters to me - all I needed from you was to let two functions into api.py, each of which seem very apropos to have since the api.py is meant to interface with misp; I'm sure that others using MISP have xml they want to upload too. 


I am too busy working on compilers and drivers at the current time to go and rescue code for a second job. I just want you to know that the PR that you've merged doesn't really help solve my problem, that it could have been done in such a way that is easier for others to adapt, and that you could have had the things you were looking *at the same time* we just kept those two functions. I totally understand that may not be organized to liking with many functions underneath the api that didn't need to be. But do you absolutely have to break our stuff?

**Comment by Rafiot** [comment]
Short answer, yes.

Few reasons:
* `keys.py`: cannot be in the repository because people we will start having API keys in PR, `csvTaxonomyFile` is not even used, `iocMispMapping` need to be moved somewhere else
* the mapping of Open IOC is partially wrong, very incomplete and I don't want to have to do support for that in the main API (in the examples is not a big deal for support purposes, and the users will more or less know what they are doing)
* adding `bs4` and `lxml` as dependency for everyone doesn't make sense: most of the people don't use XML formats with MISP anyway and if they do, it will be with stix, which is also in the tools
* `create_json_attribute` is superseded by `MISPEvent.add_attribute`
* all the OpenIOC functions are replaced by `tools.openioc.load_openioc` (+ `add_event` for upload)
* The way you implemented the XML MISP import/loading is misleading and will cause problems with other (less knowledgeable) users of the API, which mean increasing support on our end. 
* Loading an XML MISP object could be done, but it needs to actually load the *whole* event (not only the attributes). It is not in my short todo-list, especially because MISP XML export might be discontinued soonish anyway. 

Tl;DR: if you don't plan to do any changes in your current implementation, you can simply continue using your current module and stick it will still work.
I really don't understand the reason you cannot export XML documents from MISP instead of JSON. If you're missing things in the Json output, please tell me.

**Comment by KennethAdamMiller** [comment]
1) I would have made the edits you mentioned, but I didn't have the time. You didn't tell me about 2 in order for me to make it.

3) is solved by moving it into a submodule. 4) wasn't what I needed; I only need to take a misp_event and an xml file or buffer and attach it to the misp event. 

5) Yeah, but the signature is different. 6) I understand the headache about support; the less code, the better. But you were already supporting this functionality as an example; it just got moved, that's all. 

7) I had no idea I was missing anything; I tried my best to manually confirm the presents of events during the process of loading them. Perhaps for all the xml files that I had, I didn't have an example of some missing field being exercised. That's the problem with rapid development - you have standards and then the actual code, and they diverge often and by a lot.

I would prefer to come to resolution through discussion before doing any coding. I have no choice over the format, or else I would just use the first one that worked or what worked best. The reason why I have no choice is because the automated tools that we use are commercial and private; I can't make them change, and the service that their software provides is critical for our operation. It makes no difference to me whether it's xml or json personally.

**Comment by Rafiot** [comment]
1. Sure, and that's cool, it took me a while to work on your PR too

2.  I never used the OpenIOC mapping before, which is one of the reason it is in the examples and not in the core of the library.

3. `pip install xyz` is fine if you need the packages (see the way I did it in the other tools)

4. It is totally valid to do it in a wrapper using pymisp instead of changing the core: your usecase is *very* specific

5. Well, yes, because the implementation is totally different.

6. Examples are only support as examples. People using them are supposed to know what they are doing. 

7. The MISP XML format is badly documented and is supposed to disappear. By loading the whole event, I mean getting the info field, date, distribution and so on.

If there a chance you can put us in touch with your vendor? so we explain them that they should at least allow their customers to get the json format, because only XML is a really bad idea and it is not going to work on the long run.

**Comment by KennethAdamMiller** [comment]
4) I don't disagree with putting it in a module, and I think I may have said that. But nevermind the little things anyway, they don't matter.

5) Right, I was much more concerned about this. But the mountain has moved because of something different you've said

7) Yeah, the commercial tool we were paying for was excellent only as a sandbox. The UI, API and the other stuff was all really poor to be honest. I have no control over what fields were populated, so I went by what the structure of the examples coming out actually were, since all I ever saw as far as MISP info goes was the attribute field used - for everything. 

Anyway, since you say that MISP xml is going to disappear that changes a lot. Their xml content is poor anyway, and I'm not sure it's worth it trying to chase them for ever little feature and bug because we can only discover the shortcomings as we go.
